main/exercise/ptest_category.class.php

```diff
@@ -0,0 +1,1174 @@
+                        false,
+                        true
+                    )
+                );
+                $row++;
+            }
+            if (!empty($total)) {
+                $table->setCellContents($row, 0, get_lang('Total'));
+                $table->setCellContents(
+                    $row,
+                    1,
+                    ExerciseLib::show_score(
+                        $total['score'],
+                        $total['total'],
+                        false
+                    )
+                );
+                $table->setCellContents(
+                    $row,
+                    2,
+                    ExerciseLib::show_score(
+                        $total['score'],
+                        $total['total'],
+                        true,
+                        false,
+                        true
+                    )
+                );
+            }
+
+            return $table->toHtml();
+        }
+
+        return '';
+    }
+
+    /**
+     * @param Exercise $exercise
+     * @param int      $courseId
+     * @param string   $order
+     * @param bool     $shuffle
+     * @param bool     $excludeCategoryWithNoQuestions
+     *
+     * @return array
+     */
+    public function getCategoryExerciseTree(
+        $exercise,
+        $courseId,
+        $order = null,
+        $shuffle = false,
+        $excludeCategoryWithNoQuestions = true
+    ) {
+        if (empty($exercise)) {
+            return [];
+        }
+
+        $courseId = (int) $courseId;
+        $table = Database::get_course_table(TABLE_QUIZ_REL_CATEGORY);
+        $categoryTable = Database::get_course_table(TABLE_QUIZ_CATEGORY_PTEST);
+        $exercise->id = (int) $exercise->id;
+
+        $sql = "SELECT * FROM $table qc
+                LEFT JOIN $categoryTable c
+                ON (qc.c_id = c.c_id AND c.id = qc.category_id)
+                WHERE qc.c_id = $courseId AND exercise_id = {$exercise->id} ";
+
+        if (!empty($order)) {
+            $order = Database::escape_string($order);
+            $sql .= "ORDER BY $order";
+        }
+
+        $categories = [];
+        $result = Database::query($sql);
+        if (Database::num_rows($result)) {
+            while ($row = Database::fetch_array($result, 'ASSOC')) {
+                if ($excludeCategoryWithNoQuestions) {
+                    if ($row['count_questions'] == 0) {
+                        continue;
+                    }
+                }
+                if (empty($row['title']) && empty($row['category_id'])) {
+                    $row['title'] = get_lang('NoCategory');
+                }
+                $categories[$row['category_id']] = $row;
+            }
+        }
+
+        if ($shuffle) {
+            shuffle_assoc($categories);
+        }
+
+        return $categories;
+    }
+
+    /**
+     * @param FormValidator $form
+     * @param string        $action
+     */
+    public function getForm(&$form, $action = 'new')
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The variable $category_parent_list is not named in camelCase.](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5018478)